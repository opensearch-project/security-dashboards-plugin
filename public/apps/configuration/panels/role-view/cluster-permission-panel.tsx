/*
 *   Copyright OpenSearch Contributors
 *
 *   Licensed under the Apache License, Version 2.0 (the "License").
 *   You may not use this file except in compliance with the License.
 *   A copy of the License is located at
 *
 *       http://www.apache.org/licenses/LICENSE-2.0
 *
 *   or in the "license" file accompanying this file. This file is distributed
 *   on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 *   express or implied. See the License for the specific language governing
 *   permissions and limitations under the License.
 */

import React from 'react';
import { EuiEmptyPrompt, EuiButton } from '@elastic/eui';
import { PanelWithHeader } from '../../utils/panel-with-header';
import { PermissionTree } from '../permission-tree';
import { ActionGroupItem, DataObject, Action } from '../../types';
import { ResourceType } from '../../../../../common';
import { buildHashUrl } from '../../utils/url-builder';
import { loadingSpinner } from '../../utils/loading-spinner-utils';
import { DocLinks } from '../../constants';

interface ClusterPermissionPanelProps {
  roleName: string;
  clusterPermissions: string[];
  actionGroups: DataObject<ActionGroupItem>;
  loading: boolean;
  isReserved: boolean;
}

export function ClusterPermissionPanel(props: ClusterPermissionPanelProps) {
  const noClusterPermissions = (
    <EuiEmptyPrompt
      title={<h3>No cluster permission</h3>}
      titleSize="s"
      actions={
        <EuiButton
          data-test-subj="addClusterPermission"
          disabled={props.isReserved}
          onClick={() => {
            window.location.href = buildHashUrl(ResourceType.roles, Action.edit, props.roleName);
          }}
        >
          Add cluster permission
        </EuiButton>
      }
    />
  );

  const headerText = 'Cluster permissions';

  return (
    <PanelWithHeader
      headerText={headerText}
      headerSubText="Cluster permissions specify how users in this role can access the cluster. You can
      specify permissions using both action groups or single permissions. An action
      group is a list of single permissions."
      helpLink={DocLinks.ClusterPermissionsDoc}
      count={props.clusterPermissions.length}
    >
      {props.loading ? (
        <div className="text-center">{loadingSpinner}</div>
      ) : props.clusterPermissions.length === 0 ? (
        noClusterPermissions
      ) : (
        <PermissionTree permissions={props.clusterPermissions} actionGroups={props.actionGroups} />
      )}
    </PanelWithHeader>
  );
}
