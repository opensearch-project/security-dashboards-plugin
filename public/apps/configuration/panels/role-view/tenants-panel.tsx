/*
 *   Copyright OpenSearch Contributors
 *
 *   Licensed under the Apache License, Version 2.0 (the "License").
 *   You may not use this file except in compliance with the License.
 *   A copy of the License is located at
 *
 *       http://www.apache.org/licenses/LICENSE-2.0
 *
 *   or in the "license" file accompanying this file. This file is distributed
 *   on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 *   express or implied. See the License for the specific language governing
 *   permissions and limitations under the License.
 */

import React, { useState, useEffect } from 'react';
import {
  EuiInMemoryTable,
  EuiLink,
  EuiText,
  EuiGlobalToastList,
  EuiBasicTableColumn,
  EuiEmptyPrompt,
  EuiButton,
} from '@elastic/eui';
import { CoreStart } from 'opensearch-dashboards/public';
import { getCurrentUser } from '../../../../utils/auth-info-utils';
import { PanelWithHeader } from '../../utils/panel-with-header';
import {
  RoleTenantPermissionView,
  RoleTenantPermissionDetail,
  ResourceType,
  Action,
} from '../../types';
import { truncatedListView, tableItemsUIProps } from '../../utils/display-utils';
import {
  fetchTenants,
  transformTenantData,
  transformRoleTenantPermissionData,
  selectTenant,
} from '../../utils/tenant-utils';
import { DocLinks, RoleViewTenantInvalidText } from '../../constants';
import { PageId } from '../../types';
import { getNavLinkById } from '../../../../services/chrome_wrapper';
import { useToastState, createUnknownErrorToast } from '../../utils/toast-utils';
import { showTableStatusMessage } from '../../utils/loading-spinner-utils';
import { buildHashUrl } from '../../utils/url-builder';

interface RoleViewTenantsPanelProps {
  roleName: string;
  tenantPermissions: RoleTenantPermissionView[];
  errorFlag: boolean;
  coreStart: CoreStart;
  loading: boolean;
  isReserved: boolean;
}

export function TenantsPanel(props: RoleViewTenantsPanelProps) {
  const [tenantPermissionDetail, setTenantPermissionDetail] = React.useState<
    RoleTenantPermissionDetail[]
  >([]);
  const [currentUsername, setCurrentUsername] = React.useState('');
  const [toasts, addToast, removeToast] = useToastState();
  const [errorFlag, setErrorFlag] = React.useState(props.errorFlag);

  React.useEffect(() => {
    const fetchData = async () => {
      try {
        const rawTenantData = await fetchTenants(props.coreStart.http);
        const processedTenantData = transformTenantData(rawTenantData);
        setTenantPermissionDetail(
          transformRoleTenantPermissionData(props.tenantPermissions, processedTenantData)
        );
        const currentUser = await getCurrentUser(props.coreStart.http);
        setCurrentUsername(currentUser);
      } catch (e) {
        console.log(e);
        setErrorFlag(true);
      }
    };

    fetchData();
  }, [props]);

  const viewDashboard = async (tenantName: string) => {
    try {
      await selectTenant(props.coreStart.http, {
        tenant: tenantName,
        username: currentUsername,
      });
      window.location.href = getNavLinkById(props.coreStart, PageId.dashboardId);
    } catch (e) {
      console.log(e);
      addToast(createUnknownErrorToast('viewDashboard', `view dashboard for ${tenantName} tenant`));
    }
  };

  const viewVisualization = async (tenantName: string) => {
    try {
      await selectTenant(props.coreStart.http, {
        tenant: tenantName,
        username: currentUsername,
      });
      window.location.href = getNavLinkById(props.coreStart, PageId.visualizationId);
    } catch (e) {
      console.log(e);
      addToast(
        createUnknownErrorToast('viewVisualization', `view visualization for ${tenantName} tenant`)
      );
    }
  };

  const columns: Array<EuiBasicTableColumn<RoleTenantPermissionDetail>> = [
    {
      field: 'tenant_patterns',
      name: 'Name',
      render: truncatedListView(tableItemsUIProps),
      truncateText: true,
      sortable: true,
    },
    {
      field: 'description',
      name: 'Description',
      truncateText: true,
    },
    {
      field: 'permissionType',
      name: 'Read/write permission',
    },
    {
      field: 'tenantValue',
      name: 'Dashboard',
      render: (tenant: string) => {
        if (tenant === RoleViewTenantInvalidText) {
          return (
            <>
              <EuiText size="s">{RoleViewTenantInvalidText}</EuiText>
            </>
          );
        }
        return (
          <>
            <EuiLink data-test-subj="view-dashboard" onClick={() => viewDashboard(tenant)}>
              View dashboard
            </EuiLink>
          </>
        );
      },
    },
    {
      field: 'tenantValue',
      name: 'Visualizations',
      render: (tenant: string) => {
        if (tenant === RoleViewTenantInvalidText) {
          return (
            <>
              <EuiText size="s">{RoleViewTenantInvalidText}</EuiText>
            </>
          );
        }
        return (
          <>
            <EuiLink data-test-subj="view-visualizations" onClick={() => viewVisualization(tenant)}>
              View visualizations
            </EuiLink>
          </>
        );
      },
    },
  ];

  const emptyListMessage = (
    <EuiEmptyPrompt
      title={<h3>No tenant permission</h3>}
      titleSize="s"
      actions={
        <EuiButton
          data-test-subj="addTenantPermission"
          disabled={props.isReserved}
          onClick={() => {
            window.location.href = buildHashUrl(ResourceType.roles, Action.edit, props.roleName);
          }}
        >
          Add tenant permission
        </EuiButton>
      }
    />
  );

  const headerText = 'Tenant permissions';
  return (
    <>
      <PanelWithHeader
        headerText={headerText}
        headerSubText="Tenants in OpenSearch Dashboards are spaces for saving index patterns, visualizations, dashboards, and other OpenSearch Dashboards objects.
        Tenants are useful for safely sharing your work with other OpenSearch Dashboards users. You can control which roles have access to a tenant and whether those roles have read or write access."
        helpLink={DocLinks.TenantPermissionsDoc}
        count={tenantPermissionDetail.length}
      >
        <EuiInMemoryTable
          data-test-subj="tenant-permission-container"
          tableLayout={'auto'}
          loading={tenantPermissionDetail === [] && !errorFlag}
          columns={columns}
          items={tenantPermissionDetail}
          sorting={{ sort: { field: 'type', direction: 'asc' } }}
          error={errorFlag ? 'Load data failed, please check console log for more detail.' : ''}
          message={showTableStatusMessage(props.loading, props.tenantPermissions, emptyListMessage)}
        />
      </PanelWithHeader>
      <EuiGlobalToastList toasts={toasts} toastLifeTimeMs={10000} dismissToast={removeToast} />
    </>
  );
}
