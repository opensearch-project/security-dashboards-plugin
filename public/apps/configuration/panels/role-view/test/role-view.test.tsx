/*
 *   Copyright OpenSearch Contributors
 *
 *   Licensed under the Apache License, Version 2.0 (the "License").
 *   You may not use this file except in compliance with the License.
 *   A copy of the License is located at
 *
 *       http://www.apache.org/licenses/LICENSE-2.0
 *
 *   or in the "license" file accompanying this file. This file is distributed
 *   on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 *   express or implied. See the License for the specific language governing
 *   permissions and limitations under the License.
 */

import React from 'react';
import { mount, shallow } from 'enzyme';
import { RoleView } from '../role-view';
import { ClusterPermissionPanel } from '../../role-view/cluster-permission-panel';
import { IndexPermissionPanel } from '../index-permission-panel';
import { TenantsPanel } from '../tenants-panel';
import { EuiTabbedContent } from '@elastic/eui';
import {
  getRoleMappingData,
  transformRoleMappingData,
  updateRoleMapping,
} from '../../../utils/role-mapping-utils';
import { fetchActionGroups } from '../../../utils/action-groups-utils';
import { getRoleDetail } from '../../../utils/role-detail-utils';
import { transformRoleIndexPermissions } from '../../../utils/index-permission-utils';
import { useDeleteConfirmState } from '../../../utils/delete-confirm-modal-utils';
import { requestDeleteRoles } from '../../../utils/role-list-utils';
import { Action, SubAction } from '../../../types';
import { ResourceType } from '../../../../../../common';
import { buildHashUrl } from '../../../utils/url-builder';
import { createUnknownErrorToast } from '../../../utils/toast-utils';

jest.mock('../../../utils/role-mapping-utils', () => ({
  getRoleMappingData: jest.fn().mockReturnValue({ backend_roles: [], hosts: [], users: [] }),
  transformRoleMappingData: jest.fn().mockReturnValue({
    userName: '',
    userType: '',
  }),
  updateRoleMapping: jest.fn(),
}));
jest.mock('../../../utils/action-groups-utils', () => ({
  fetchActionGroups: jest.fn().mockReturnValue({}),
}));
jest.mock('../../../utils/role-detail-utils', () => ({
  getRoleDetail: jest.fn().mockReturnValue({
    cluster_permissions: [],
    index_permissions: [],
    tenant_permissions: [],
    reserved: false,
  }),
}));
jest.mock('../../../utils/delete-confirm-modal-utils', () => ({
  useDeleteConfirmState: jest.fn().mockReturnValue([jest.fn(), '']),
}));
jest.mock('../../../utils/index-permission-utils');
jest.mock('../../../utils/tenant-utils');
jest.mock('../../../utils/role-list-utils', () => ({
  requestDeleteRoles: jest.fn(),
}));
jest.mock('../../../utils/context-menu', () => ({
  useContextMenuState: jest
    .fn()
    .mockImplementation((buttonText, buttonProps, children) => [children, jest.fn()]),
}));
jest.mock('../../../utils/toast-utils', () => ({
  createErrorToast: jest.fn(),
  createUnknownErrorToast: jest.fn(),
  useToastState: jest.fn().mockReturnValue([[], jest.fn(), jest.fn()]),
}));

describe('Role view', () => {
  const setState = jest.fn();
  const sampleRole = 'role';
  const mockCoreStart = {
    http: 1,
  };
  const buildBreadcrumbs = jest.fn();

  const useEffect = jest.spyOn(React, 'useEffect');
  const useState = jest.spyOn(React, 'useState');

  beforeEach(() => {
    useEffect.mockImplementationOnce((f) => f());
    useState.mockImplementation((initialValue) => [initialValue, setState]);
  });

  it('basic rendering when permission tab is selected', () => {
    const component = shallow(
      <RoleView
        roleName={sampleRole}
        prevAction=""
        buildBreadcrumbs={buildBreadcrumbs}
        coreStart={mockCoreStart as any}
        navigation={{} as any}
        params={{} as any}
        config={{} as any}
      />
    );

    expect(buildBreadcrumbs).toBeCalledTimes(1);
    expect(component.find(EuiTabbedContent).length).toBe(1);

    const tabs = component.find(EuiTabbedContent).dive();
    expect(tabs.find(ClusterPermissionPanel).length).toBe(1);
    expect(tabs.find(IndexPermissionPanel).length).toBe(1);
    expect(tabs.find(TenantsPanel).length).toBe(1);
    expect(component).toMatchSnapshot();
  });

  it('renders when mapped user tab is selected', () => {
    const component = shallow(
      <RoleView
        roleName={sampleRole}
        prevAction={SubAction.mapuser}
        buildBreadcrumbs={buildBreadcrumbs}
        coreStart={mockCoreStart as any}
        navigation={{} as any}
        params={{} as any}
        config={{} as any}
      />
    );
    expect(component).toMatchSnapshot();
  });

  it('should render to map user page when click on Map users', () => {
    const wrapper = shallow(
      <RoleView
        roleName={sampleRole}
        prevAction={SubAction.mapuser}
        buildBreadcrumbs={buildBreadcrumbs}
        coreStart={mockCoreStart as any}
        navigation={{} as any}
        params={{} as any}
        config={{} as any}
      />
    );
    const tabs = wrapper.find(EuiTabbedContent).dive();
    const roleMappingList = tabs.find('[data-test-subj="role-mapping-list"]').dive();
    const Wrapper = mount(<>{roleMappingList}</>);
    Wrapper.find('[data-test-subj="map-users"]').first().simulate('click');
    expect(window.location.hash).toBe(
      buildHashUrl(ResourceType.roles, Action.edit, sampleRole, SubAction.mapuser)
    );
  });

  it('should render to map user page when click on Manage Mapping', () => {
    const wrapper = shallow(
      <RoleView
        roleName={sampleRole}
        prevAction={SubAction.mapuser}
        buildBreadcrumbs={buildBreadcrumbs}
        coreStart={mockCoreStart as any}
        navigation={{} as any}
        params={{} as any}
        config={{} as any}
      />
    );
    const tabs = wrapper.find(EuiTabbedContent).dive();
    tabs.find('[data-test-subj="manage-mapping"]').first().simulate('click');
    expect(window.location.hash).toBe(
      buildHashUrl(ResourceType.roles, Action.edit, sampleRole, SubAction.mapuser)
    );
  });

  it('fetch data', (done) => {
    const component = shallow(
      <RoleView
        roleName={sampleRole}
        prevAction=""
        buildBreadcrumbs={buildBreadcrumbs}
        coreStart={mockCoreStart as any}
        navigation={{} as any}
        params={{} as any}
        config={{} as any}
      />
    );

    process.nextTick(() => {
      expect(getRoleMappingData).toHaveBeenCalledTimes(1);
      expect(transformRoleMappingData).toHaveBeenCalledTimes(1);
      expect(fetchActionGroups).toHaveBeenCalledTimes(1);
      expect(getRoleDetail).toHaveBeenCalledTimes(1);
      expect(transformRoleIndexPermissions).toHaveBeenCalledTimes(1);
      done();
    });
  });

  it('fetch data error', () => {
    getRoleMappingData.mockImplementationOnce(() => {
      throw new Error();
    });
    // Hide the error message
    jest.spyOn(console, 'log').mockImplementationOnce(() => {});

    shallow(
      <RoleView
        roleName={sampleRole}
        prevAction=""
        buildBreadcrumbs={buildBreadcrumbs}
        coreStart={mockCoreStart as any}
        navigation={{} as any}
        params={{} as any}
        config={{} as any}
      />
    );

    expect(setState).toHaveBeenCalledWith(true);
  });

  it('delete role mapping', (done) => {
    shallow(
      <RoleView
        roleName={sampleRole}
        prevAction=""
        buildBreadcrumbs={buildBreadcrumbs}
        coreStart={mockCoreStart as any}
        navigation={{} as any}
        params={{} as any}
        config={{} as any}
      />
    );
    const deleteFunc = useDeleteConfirmState.mock.calls[0][0];

    deleteFunc();

    process.nextTick(() => {
      expect(updateRoleMapping).toBeCalled();
      done();
    });
  });

  it('should capture error by console.log if error occurred while deleting role mapping', (done) => {
    (updateRoleMapping as jest.Mock).mockImplementationOnce(() => {
      throw new Error();
    });
    const spy = jest.spyOn(console, 'log').mockImplementationOnce(() => {});
    shallow(
      <RoleView
        roleName={sampleRole}
        prevAction=""
        buildBreadcrumbs={buildBreadcrumbs}
        coreStart={mockCoreStart as any}
        navigation={{} as any}
        params={{} as any}
        config={{} as any}
      />
    );
    const deleteFunc = useDeleteConfirmState.mock.calls[0][0];

    deleteFunc();

    process.nextTick(() => {
      expect(spy).toBeCalled();
      done();
    });
  });

  it('delete role', () => {
    const component = shallow(
      <RoleView
        roleName={sampleRole}
        prevAction=""
        buildBreadcrumbs={buildBreadcrumbs}
        coreStart={mockCoreStart as any}
        navigation={{} as any}
        params={{} as any}
        config={{} as any}
      />
    );
    component.find('[data-test-subj="delete"]').simulate('click');

    expect(requestDeleteRoles).toBeCalled();
  });

  it('error occurred while deleting the role', () => {
    (requestDeleteRoles as jest.Mock).mockImplementationOnce(() => {
      throw new Error();
    });
    const component = shallow(
      <RoleView
        roleName={sampleRole}
        prevAction=""
        buildBreadcrumbs={buildBreadcrumbs}
        coreStart={mockCoreStart as any}
        navigation={{} as any}
        params={{} as any}
        config={{} as any}
      />
    );
    component.find('[data-test-subj="delete"]').simulate('click');
    expect(createUnknownErrorToast).toBeCalled();
  });
});
