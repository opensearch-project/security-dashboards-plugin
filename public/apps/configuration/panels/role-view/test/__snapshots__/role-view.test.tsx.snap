// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Role view basic rendering when permission tab is selected 1`] = `
<Fragment>
  <EuiPageContentHeader>
    <EuiPageContentHeaderSection>
      <EuiTitle
        size="l"
      >
        <h1>
          role
        </h1>
      </EuiTitle>
    </EuiPageContentHeaderSection>
    <EuiPageContentHeaderSection>
      <EuiFlexGroup
        gutterSize="s"
      >
        <EuiFlexItem>
          <EuiButtonEmpty
            href="#/roles/duplicate/role"
            key="duplicate"
          >
            duplicate
          </EuiButtonEmpty>
          <EuiButtonEmpty
            color="danger"
            data-test-subj="delete"
            key="delete"
            onClick={[Function]}
          >
            delete
          </EuiButtonEmpty>
        </EuiFlexItem>
        <EuiFlexItem>
          <EuiButton
            href="#/roles/edit/role"
          >
            Edit role
          </EuiButton>
        </EuiFlexItem>
      </EuiFlexGroup>
    </EuiPageContentHeaderSection>
  </EuiPageContentHeader>
  <EuiTabbedContent
    autoFocus="initial"
    initialSelectedTab={
      Object {
        "content": <React.Fragment>
          <EuiSpacer
            size="m"
          />
          <EuiSpacer
            size="m"
          />
          <ClusterPermissionPanel
            actionGroups={Object {}}
            clusterPermissions={Array []}
            isReserved={false}
            loading={false}
            roleName="role"
          />
          <EuiSpacer
            size="m"
          />
          <IndexPermissionPanel
            actionGroups={Object {}}
            errorFlag={false}
            indexPermissions={Array []}
            isReserved={false}
            loading={false}
            roleName="role"
          />
          <EuiSpacer
            size="m"
          />
          <TenantsPanel
            coreStart={
              Object {
                "http": 1,
              }
            }
            errorFlag={false}
            isReserved={false}
            loading={false}
            roleName="role"
            tenantPermissions={Array []}
          />
        </React.Fragment>,
        "disabled": false,
        "id": "permissions",
        "name": "Permissions",
      }
    }
    tabs={
      Array [
        Object {
          "content": <React.Fragment>
            <EuiSpacer
              size="m"
            />
            <EuiSpacer
              size="m"
            />
            <ClusterPermissionPanel
              actionGroups={Object {}}
              clusterPermissions={Array []}
              isReserved={false}
              loading={false}
              roleName="role"
            />
            <EuiSpacer
              size="m"
            />
            <IndexPermissionPanel
              actionGroups={Object {}}
              errorFlag={false}
              indexPermissions={Array []}
              isReserved={false}
              loading={false}
              roleName="role"
            />
            <EuiSpacer
              size="m"
            />
            <TenantsPanel
              coreStart={
                Object {
                  "http": 1,
                }
              }
              errorFlag={false}
              isReserved={false}
              loading={false}
              roleName="role"
              tenantPermissions={Array []}
            />
          </React.Fragment>,
          "disabled": false,
          "id": "permissions",
          "name": "Permissions",
        },
        Object {
          "content": <React.Fragment>
            <EuiSpacer />
            <EuiPageContent>
              <EuiPageContentHeader>
                <EuiPageContentHeaderSection>
                  <EuiTitle
                    size="s"
                  >
                    <h3>
                      Mapped users
                      <span
                        className="panel-header-count"
                      >
                         (
                        0
                        )
                      </span>
                    </h3>
                  </EuiTitle>
                  <EuiText
                    className="panel-header-subtext"
                    color="subdued"
                    size="xs"
                  >
                    You can map two types of users: users and backend roles. A user can have its own backend role and host for an external authentication and authorization. A backend role directly maps to roles through an external authentication system.
                     
                    <ExternalLink
                      href="https://opensearch.org/docs/latest/security-plugin/access-control/users-roles/#map-users-to-roles"
                    />
                  </EuiText>
                </EuiPageContentHeaderSection>
                <EuiPageContentHeaderSection>
                  <EuiFlexGroup>
                    <EuiFlexItem>
                      <EuiButton
                        disabled={true}
                        onClick={[MockFunction]}
                      >
                        Delete mapping
                      </EuiButton>
                    </EuiFlexItem>
                    <EuiFlexItem>
                      <EuiButton
                        data-test-subj="manage-mapping"
                        onClick={[Function]}
                      >
                        Manage mapping
                      </EuiButton>
                    </EuiFlexItem>
                  </EuiFlexGroup>
                </EuiPageContentHeaderSection>
              </EuiPageContentHeader>
              <EuiHorizontalRule
                margin="s"
              />
              <EuiPageBody>
                <EuiInMemoryTable
                  columns={
                    Array [
                      Object {
                        "field": "userType",
                        "name": "User type",
                        "sortable": true,
                      },
                      Object {
                        "field": "userName",
                        "name": "User",
                        "sortable": true,
                        "truncateText": true,
                      },
                    ]
                  }
                  data-test-subj="role-mapping-list"
                  error=""
                  itemId="userName"
                  items={Array []}
                  loading={false}
                  message={
                    <EuiEmptyPrompt
                      actions={
                        <EuiFlexGroup
                          alignItems="center"
                          gutterSize="s"
                        >
                          <EuiFlexItem>
                            <ExternalLinkButton
                              href="#/users/create"
                              text="Create internal user"
                            />
                          </EuiFlexItem>
                          <EuiFlexItem>
                            <EuiButton
                              data-test-subj="map-users"
                              fill={true}
                              onClick={[Function]}
                            >
                              Map users
                            </EuiButton>
                          </EuiFlexItem>
                        </EuiFlexGroup>
                      }
                      body={
                        <EuiText
                          color="subdued"
                          grow={false}
                          size="s"
                        >
                          <p>
                            You can map users or backend roles to this role
                          </p>
                        </EuiText>
                      }
                      title={
                        <h2>
                          No user has been mapped to this role
                        </h2>
                      }
                      titleSize="s"
                    />
                  }
                  pagination={true}
                  responsive={true}
                  selection={
                    Object {
                      "onSelectionChange": [Function],
                    }
                  }
                  sorting={true}
                  tableLayout="auto"
                />
              </EuiPageBody>
            </EuiPageContent>
          </React.Fragment>,
          "disabled": false,
          "id": "users",
          "name": "Mapped users",
        },
      ]
    }
  />
  <EuiSpacer />
  <EuiGlobalToastList
    dismissToast={[MockFunction]}
    side="right"
    toastLifeTimeMs={10000}
    toasts={Array []}
  />
</Fragment>
`;

exports[`Role view renders when mapped user tab is selected 1`] = `
<Fragment>
  <EuiPageContentHeader>
    <EuiPageContentHeaderSection>
      <EuiTitle
        size="l"
      >
        <h1>
          role
        </h1>
      </EuiTitle>
    </EuiPageContentHeaderSection>
    <EuiPageContentHeaderSection>
      <EuiFlexGroup
        gutterSize="s"
      >
        <EuiFlexItem>
          <EuiButtonEmpty
            href="#/roles/duplicate/role"
            key="duplicate"
          >
            duplicate
          </EuiButtonEmpty>
          <EuiButtonEmpty
            color="danger"
            data-test-subj="delete"
            key="delete"
            onClick={[Function]}
          >
            delete
          </EuiButtonEmpty>
        </EuiFlexItem>
        <EuiFlexItem>
          <EuiButton
            href="#/roles/edit/role"
          >
            Edit role
          </EuiButton>
        </EuiFlexItem>
      </EuiFlexGroup>
    </EuiPageContentHeaderSection>
  </EuiPageContentHeader>
  <EuiTabbedContent
    autoFocus="initial"
    initialSelectedTab={
      Object {
        "content": <React.Fragment>
          <EuiSpacer />
          <EuiPageContent>
            <EuiPageContentHeader>
              <EuiPageContentHeaderSection>
                <EuiTitle
                  size="s"
                >
                  <h3>
                    Mapped users
                    <span
                      className="panel-header-count"
                    >
                       (
                      0
                      )
                    </span>
                  </h3>
                </EuiTitle>
                <EuiText
                  className="panel-header-subtext"
                  color="subdued"
                  size="xs"
                >
                  You can map two types of users: users and backend roles. A user can have its own backend role and host for an external authentication and authorization. A backend role directly maps to roles through an external authentication system.
                   
                  <ExternalLink
                    href="https://opensearch.org/docs/latest/security-plugin/access-control/users-roles/#map-users-to-roles"
                  />
                </EuiText>
              </EuiPageContentHeaderSection>
              <EuiPageContentHeaderSection>
                <EuiFlexGroup>
                  <EuiFlexItem>
                    <EuiButton
                      disabled={true}
                      onClick={[MockFunction]}
                    >
                      Delete mapping
                    </EuiButton>
                  </EuiFlexItem>
                  <EuiFlexItem>
                    <EuiButton
                      data-test-subj="manage-mapping"
                      onClick={[Function]}
                    >
                      Manage mapping
                    </EuiButton>
                  </EuiFlexItem>
                </EuiFlexGroup>
              </EuiPageContentHeaderSection>
            </EuiPageContentHeader>
            <EuiHorizontalRule
              margin="s"
            />
            <EuiPageBody>
              <EuiInMemoryTable
                columns={
                  Array [
                    Object {
                      "field": "userType",
                      "name": "User type",
                      "sortable": true,
                    },
                    Object {
                      "field": "userName",
                      "name": "User",
                      "sortable": true,
                      "truncateText": true,
                    },
                  ]
                }
                data-test-subj="role-mapping-list"
                error=""
                itemId="userName"
                items={Array []}
                loading={false}
                message={
                  <EuiEmptyPrompt
                    actions={
                      <EuiFlexGroup
                        alignItems="center"
                        gutterSize="s"
                      >
                        <EuiFlexItem>
                          <ExternalLinkButton
                            href="#/users/create"
                            text="Create internal user"
                          />
                        </EuiFlexItem>
                        <EuiFlexItem>
                          <EuiButton
                            data-test-subj="map-users"
                            fill={true}
                            onClick={[Function]}
                          >
                            Map users
                          </EuiButton>
                        </EuiFlexItem>
                      </EuiFlexGroup>
                    }
                    body={
                      <EuiText
                        color="subdued"
                        grow={false}
                        size="s"
                      >
                        <p>
                          You can map users or backend roles to this role
                        </p>
                      </EuiText>
                    }
                    title={
                      <h2>
                        No user has been mapped to this role
                      </h2>
                    }
                    titleSize="s"
                  />
                }
                pagination={true}
                responsive={true}
                selection={
                  Object {
                    "onSelectionChange": [Function],
                  }
                }
                sorting={true}
                tableLayout="auto"
              />
            </EuiPageBody>
          </EuiPageContent>
        </React.Fragment>,
        "disabled": false,
        "id": "users",
        "name": "Mapped users",
      }
    }
    tabs={
      Array [
        Object {
          "content": <React.Fragment>
            <EuiSpacer
              size="m"
            />
            <EuiSpacer
              size="m"
            />
            <ClusterPermissionPanel
              actionGroups={Object {}}
              clusterPermissions={Array []}
              isReserved={false}
              loading={false}
              roleName="role"
            />
            <EuiSpacer
              size="m"
            />
            <IndexPermissionPanel
              actionGroups={Object {}}
              errorFlag={false}
              indexPermissions={Array []}
              isReserved={false}
              loading={false}
              roleName="role"
            />
            <EuiSpacer
              size="m"
            />
            <TenantsPanel
              coreStart={
                Object {
                  "http": 1,
                }
              }
              errorFlag={false}
              isReserved={false}
              loading={false}
              roleName="role"
              tenantPermissions={Array []}
            />
          </React.Fragment>,
          "disabled": false,
          "id": "permissions",
          "name": "Permissions",
        },
        Object {
          "content": <React.Fragment>
            <EuiSpacer />
            <EuiPageContent>
              <EuiPageContentHeader>
                <EuiPageContentHeaderSection>
                  <EuiTitle
                    size="s"
                  >
                    <h3>
                      Mapped users
                      <span
                        className="panel-header-count"
                      >
                         (
                        0
                        )
                      </span>
                    </h3>
                  </EuiTitle>
                  <EuiText
                    className="panel-header-subtext"
                    color="subdued"
                    size="xs"
                  >
                    You can map two types of users: users and backend roles. A user can have its own backend role and host for an external authentication and authorization. A backend role directly maps to roles through an external authentication system.
                     
                    <ExternalLink
                      href="https://opensearch.org/docs/latest/security-plugin/access-control/users-roles/#map-users-to-roles"
                    />
                  </EuiText>
                </EuiPageContentHeaderSection>
                <EuiPageContentHeaderSection>
                  <EuiFlexGroup>
                    <EuiFlexItem>
                      <EuiButton
                        disabled={true}
                        onClick={[MockFunction]}
                      >
                        Delete mapping
                      </EuiButton>
                    </EuiFlexItem>
                    <EuiFlexItem>
                      <EuiButton
                        data-test-subj="manage-mapping"
                        onClick={[Function]}
                      >
                        Manage mapping
                      </EuiButton>
                    </EuiFlexItem>
                  </EuiFlexGroup>
                </EuiPageContentHeaderSection>
              </EuiPageContentHeader>
              <EuiHorizontalRule
                margin="s"
              />
              <EuiPageBody>
                <EuiInMemoryTable
                  columns={
                    Array [
                      Object {
                        "field": "userType",
                        "name": "User type",
                        "sortable": true,
                      },
                      Object {
                        "field": "userName",
                        "name": "User",
                        "sortable": true,
                        "truncateText": true,
                      },
                    ]
                  }
                  data-test-subj="role-mapping-list"
                  error=""
                  itemId="userName"
                  items={Array []}
                  loading={false}
                  message={
                    <EuiEmptyPrompt
                      actions={
                        <EuiFlexGroup
                          alignItems="center"
                          gutterSize="s"
                        >
                          <EuiFlexItem>
                            <ExternalLinkButton
                              href="#/users/create"
                              text="Create internal user"
                            />
                          </EuiFlexItem>
                          <EuiFlexItem>
                            <EuiButton
                              data-test-subj="map-users"
                              fill={true}
                              onClick={[Function]}
                            >
                              Map users
                            </EuiButton>
                          </EuiFlexItem>
                        </EuiFlexGroup>
                      }
                      body={
                        <EuiText
                          color="subdued"
                          grow={false}
                          size="s"
                        >
                          <p>
                            You can map users or backend roles to this role
                          </p>
                        </EuiText>
                      }
                      title={
                        <h2>
                          No user has been mapped to this role
                        </h2>
                      }
                      titleSize="s"
                    />
                  }
                  pagination={true}
                  responsive={true}
                  selection={
                    Object {
                      "onSelectionChange": [Function],
                    }
                  }
                  sorting={true}
                  tableLayout="auto"
                />
              </EuiPageBody>
            </EuiPageContent>
          </React.Fragment>,
          "disabled": false,
          "id": "users",
          "name": "Mapped users",
        },
      ]
    }
  />
  <EuiSpacer />
  <EuiGlobalToastList
    dismissToast={[MockFunction]}
    side="right"
    toastLifeTimeMs={10000}
    toasts={Array []}
  />
</Fragment>
`;
